### PR TITLE
implement top_of_window and bottom_of_window

### PR DESCRIPTION
**Pull request by zhaojiangbin**
to complement center_of_window

### CODE REVIEW

**Comment by zhaojiangbin** [comment]
> @zhaojiangbin Could you please rebase this?

It is now rebased.

**Review by bugadani — APPROVED** [review]

File: lapce-ui/src/editor/view.rs [hunk 1]
```diff
@@ -259,19 +259,51 @@ impl LapceEditorView {
...
+                let h = if h > margin { h - margin } else { margin };
+                Point::new(cursor.x, cursor.y + h as f64)
+            },
+            EnsureVisiblePosition::BottomOfWindow => {
+                // Cursor line will be at the bottom edge, thus center of the
+                // view is half height up.
+                let h = height as usize;
+                let h = if h > margin { h - margin } else { margin };
+                let y = cursor.y as usize;
+                let y = if h > y { y } else { y - h };
```
Comment (bugadani) [inline lapce-ui/src/editor/view.rs:302]: This pattern is repeated often enough to deserve its own function. This is basically `saturating_sub`, right? Also I think this code is complicated enough to warrant some tests.

**Comment by zhaojiangbin** [comment]
@dzhou121, what do you think of this PR?

**Comment by dzhou121** [comment]
I assume this is for ```zt``` and ```zb```, can you please add the default key bindings as well?

File: lapce-ui/src/editor/view.rs (hunk 1 above)
Comment (zhaojiangbin) [reply lapce-ui/src/editor/view.rs]: > This pattern is repeated often enough to deserve its own function. ~This is basically `saturating_sub`, right?~ No it's not. 

Indeed the calculation was incorrect. Fixed in the later commits.

> Anyway, I also think this code is complicated enough to warrant some tests.

I assume that you meant unit tests? I was trying to add some but could not figure out how to verify it without repeating the same calculation in the tests. I need to think about it.

**Comment by zhaojiangbin** [comment]
> I assume this is for `zt` and `zb`, can you please add the default key bindings as well?

Done.

**Comment by dzhou121** [comment]
Thanks!

**Pull request merged**